File: src/main/java/itmo/is/project/controller/module/DockModuleRestController.java
package itmo.is.project.controller.module;

import itmo.is.project.dto.module.BuildModuleRequest;
import itmo.is.project.dto.module.dock.DockModuleBlueprintDto;
import itmo.is.project.dto.module.dock.DockModuleDto;
import itmo.is.project.dto.module.dock.DockingSpotDto;
import itmo.is.project.model.user.User;
import itmo.is.project.service.module.DockModuleService;
import itmo.is.project.service.module.build.DockModuleBuildService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/modules/dock")
@RequiredArgsConstructor
public class DockModuleRestController {
    private final DockModuleService dockModuleService;
    private final DockModuleBuildService dockModuleBuildService;

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping
    public ResponseEntity<Page<DockModuleDto>> getAllDockModules(@PageableDefault Pageable pageable) {
        return ResponseEntity.ok(dockModuleService.getAllDockModules(pageable));
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/{dockModuleId}")
    public ResponseEntity<DockModuleDto> getDockModuleById(@PathVariable Integer dockModuleId) {
        return ResponseEntity.ok(dockModuleService.getDockModuleById(dockModuleId));
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/build/blueprints")
    public ResponseEntity<Page<DockModuleBlueprintDto>> findAllBlueprints(@PageableDefault Pageable pageable) {
        return ResponseEntity.ok(dockModuleBuildService.findAllBlueprints(pageable));
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @PostMapping("/build")
    public ResponseEntity<DockModuleDto> buildModule(@RequestBody BuildModuleRequest buildModuleRequest) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(dockModuleBuildService.buildModule(buildModuleRequest));
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/spots")
    public ResponseEntity<Page<DockingSpotDto>> getAllDockingSpots(@PageableDefault Pageable pageable) {
        return ResponseEntity.ok(dockModuleService.getAllDockingSpots(pageable));
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/spots/occupied")
    public ResponseEntity<Page<DockingSpotDto>> getAllOccupiedDockingSpots(@PageableDefault Pageable pageable) {
        return ResponseEntity.ok(dockModuleService.getAllOccupiedDockingSpots(pageable));
    }

    @GetMapping("/spots/occupied/my")
    public ResponseEntity<DockingSpotDto> getMyDockingSpot(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(dockModuleService.getOccupiedDockingSpotByPilot(user));
    }

    @PostMapping("/requests/dock")
    public ResponseEntity<DockingSpotDto> dock(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(dockModuleService.dock(user));
    }

    @PostMapping("/requests/undock")
    public ResponseEntity<Void> undock(@AuthenticationPrincipal User user) {
        dockModuleService.undock(user);
        return ResponseEntity.ok().build();
    }
}

File: src/main/kotlin/itmo/isproject/controller/module/DockModuleRestController.kt
package itmo.isproject.controller.module

import itmo.isproject.dto.module.BuildModuleRequest
import itmo.isproject.dto.module.dock.DockModuleBlueprintDto
import itmo.isproject.dto.module.dock.DockModuleDto
import itmo.isproject.dto.module.dock.DockingSpotDto
import itmo.isproject.model.user.User
import itmo.isproject.service.module.DockModuleService
import itmo.isproject.service.module.build.DockModuleBuildService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/modules/dock")
class DockModuleRestController(
    private val dockModuleService: DockModuleService,
    private val dockModuleBuildService: DockModuleBuildService
) {

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping
    fun getAllDockModules(@PageableDefault pageable: Pageable): ResponseEntity<Page<DockModuleDto>> {
        return ResponseEntity.ok(dockModuleService.getAllDockModules(pageable))
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/{dockModuleId}")
    fun getDockModuleById(@PathVariable dockModuleId: Int): ResponseEntity<DockModuleDto> {
        return ResponseEntity.ok(dockModuleService.getDockModuleById(dockModuleId))
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/build/blueprints")
    fun findAllBlueprints(@PageableDefault pageable: Pageable): ResponseEntity<Page<DockModuleBlueprintDto>> {
        return ResponseEntity.ok(dockModuleBuildService.findAllBlueprints(pageable))
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @PostMapping("/build")
    fun buildModule(@RequestBody buildModuleRequest: BuildModuleRequest): ResponseEntity<DockModuleDto> {
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(dockModuleBuildService.buildModule(buildModuleRequest))
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/spots")
    fun getAllDockingSpots(@PageableDefault pageable: Pageable): ResponseEntity<Page<DockingSpotDto>> {
        return ResponseEntity.ok(dockModuleService.getAllDockingSpots(pageable))
    }

    @PreAuthorize("hasRole('ROLE_OWNER') or hasRole('ROLE_MANAGER')")
    @GetMapping("/spots/occupied")
    fun getAllOccupiedDockingSpots(@PageableDefault pageable: Pageable): ResponseEntity<Page<DockingSpotDto>> {
        return ResponseEntity.ok(dockModuleService.getAllOccupiedDockingSpots(pageable))
    }

    @GetMapping("/spots/occupied/my")
    fun getMyDockingSpot(@AuthenticationPrincipal user: User): ResponseEntity<DockingSpotDto> {
        return ResponseEntity.ok(dockModuleService.getOccupiedDockingSpotByPilot(user))
    }

    @PostMapping("/requests/dock")
    fun dock(@AuthenticationPrincipal user: User): ResponseEntity<DockingSpotDto> {
        return ResponseEntity.ok(dockModuleService.dock(user))
    }

    @PostMapping("/requests/undock")
    fun undock(@AuthenticationPrincipal user: User): ResponseEntity<Void> {
        dockModuleService.undock(user)
        return ResponseEntity.ok().build()
    }
}
